solarforecastarbiter/metrics/calculator.py

```diff
@@ -149,6 +150,15 @@ def calculate_deterministic_metrics(processed_fx_obs, categories, metrics,
     elif len(metrics) == 0:
         raise RuntimeError("No metrics specified.")
 
+    # Force `groupby` to be consistent with `interval_label`, i.e., if
+    # `interval_label == ending`, then the last interval should be in the bin
+    closed = datamodel.CLOSED_MAPPING[processed_fx_obs.interval_label]
+    if closed == "ending":
+        obs = copy.deepcopy(obs)
+        fx = copy.deepcopy(fx)
+        obs.index -= pd.Timedelta("1ns")
+        fx.index -= pd.Timedelta("1ns")
+
```

Comment:
shouldn't the shift also be applied to `ref_fx`? Perhaps it could could be applied to `df.index` and we could avoid the `deepcopy` of the entire object. All of this assumes that `obs`, `fx`, and `ref_fx` already have been aligned to use the same `interval_label`.

Reply:
Good catch (I forgot about the reference forecast). And yes, applying to `df.index` does make more sense. I'll revise accordingly.

Reply:
Revised to apply the 1 ns shift to `df.index`. Also, since the input to this function is `processed_fx_obs`, I take that has meaning `resample_and_align()` has taken care of all alignment issues (although we may still need revisions to `resample_and_align()` for the `interval_label` issue we've discussed).

Reply:
We should document the requirement for the two (test fx + obs, ref fx + obs) different parameters to have the same interval label. Maybe check and raise a `ValueError` if not. We can address in a separate issue if you prefer.

Reply:
I agree with documenting and raising a `ValueError` on the interval label mismatch. (Although this may be temporary as we revise the `resample_and_align()` code.)

Reply:
**Update:** the `obs` and `fx` in the `calculate_deterministic_metrics()` code are `pd.Series` objects stored inside a `ProcessedForecastObservation` object and therefore do not have `interval_label` attributes. (Whereas the `ProcessedForecastObservation` object does have a single `interval_label` for both the `obs` and `fx`.)

So the only thing we can check directly is that the `interval_label` of the `obs`/`fx` matches to the `ref_fx`.